src/main/java/org/cryptomator/linux/tray/AppindicatorTrayMenuController.java

```diff
@@ -0,0 +1,110 @@
+package org.cryptomator.linux.tray;
+
+import org.apache.commons.lang3.SystemUtils;
+import org.cryptomator.integrations.common.CheckAvailability;
+import org.cryptomator.integrations.tray.ActionItem;
+import org.cryptomator.integrations.tray.SeparatorItem;
+import org.cryptomator.integrations.tray.SubMenuItem;
+import org.cryptomator.integrations.tray.TrayMenuController;
+import org.cryptomator.integrations.tray.TrayMenuException;
+import org.cryptomator.integrations.tray.TrayMenuItem;
+import org.purejava.appindicator.MemoryAllocator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.SegmentScope;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Paths;
+import java.util.List;
+
+import static org.purejava.appindicator.app_indicator_h.*;
+
+@CheckAvailability
+public class AppindicatorTrayMenuController implements TrayMenuController {
+
+	private static final Logger LOG = LoggerFactory.getLogger(AppindicatorTrayMenuController.class);
+
+	private final SegmentScope scope = SegmentScope.auto();
+	private MemorySegment indicator;
+	private MemorySegment menu = gtk_menu_new();
+
+	@CheckAvailability
+	public static boolean isAvailable() {
+		return SystemUtils.IS_OS_LINUX && MemoryAllocator.isLoadedNativeLib();
+	}
+
+	@Override
+	public void showTrayIcon(URI uri, Runnable runnable, String s) throws TrayMenuException {
+		indicator = app_indicator_new(MemoryAllocator.ALLOCATE_FOR("org.cryptomator.Cryptomator"),
+				MemoryAllocator.ALLOCATE_FOR(getAbsolutePath(getPathString(uri))),
+				APP_INDICATOR_CATEGORY_APPLICATION_STATUS());
+		gtk_widget_show_all(menu);
+		app_indicator_set_menu(indicator, menu);
+		app_indicator_set_status(indicator, APP_INDICATOR_STATUS_ACTIVE());
+	}
+
+	@Override
+	public void updateTrayIcon(URI uri) {
+		app_indicator_set_icon(indicator, MemoryAllocator.ALLOCATE_FOR(getAbsolutePath(getPathString(uri))));
+	}
+
+	@Override
+	public void updateTrayMenu(List<TrayMenuItem> items) throws TrayMenuException {
+		menu = gtk_menu_new();
+		addChildren(menu, items);
+		gtk_widget_show_all(menu);
+		app_indicator_set_menu(indicator, menu);
+	}
+
+	@Override
+	public void onBeforeOpenMenu(Runnable runnable) {
+
+	}
+
+	private void addChildren(MemorySegment menu, List<TrayMenuItem> items) {
+		for (var item : items) {
+			// TODO: use Pattern Matching for switch, once available
+			if (item instanceof ActionItem a) {
+				var gtkMenuItem = gtk_menu_item_new();
+				gtk_menu_item_set_label(gtkMenuItem, MemoryAllocator.ALLOCATE_FOR(a.title()));
+				g_signal_connect_object(gtkMenuItem,
+						MemoryAllocator.ALLOCATE_FOR("activate"),
+						MemoryAllocator.ALLOCATE_CALLBACK_FOR(new ActionItemCallback(a), scope),
```

Comment:
Will the callback still work after garbage collecting? I have never used `SegmentScope.auto`, so I can't speak out of experience. But I would prefer explicitly opening an [Arena](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/lang/foreign/Arena.html) and using it's scope instead.

The arena can also be used as an allocator for utf8 c strings btw.

[This article by Panama's lead dev Maurizio Cimadamore](https://cr.openjdk.org/~mcimadamore/panama/why_lifetimes.html#enter-lifetimes) is a good read on the concept of the different lifecycles.

Reply:
> Will the callback still work after garbage collecting? I have never used `SegmentScope.auto`, so I can't speak out of experience. But I would prefer explicitly opening an [Arena](https://docs.oracle.com/en/java/javase/20/docs/api/java.base/java/lang/foreign/Arena.html) and using it's scope instead.
> 

It does work. Otherwise the JVM crashes, when trying to access a callback, that has been gc'ed. The callback does work here without flaws.

Besides that, `jextract` generates a `SegmentScope`:
```java
public interface GCallback {

    void apply();
    static MemorySegment allocate(GCallback fi, SegmentScope scope) {
        return RuntimeHelper.upcallStub(constants$0.GCallback_UP$MH, fi, constants$0.GCallback$FUNC, scope);
    }
    static GCallback ofAddress(MemorySegment addr, SegmentScope scope) {
        MemorySegment symbol = MemorySegment.ofAddress(addr.address(), 0, scope);
        return () -> {
            try {
                constants$0.GCallback_DOWN$MH.invokeExact(symbol);
            } catch (Throwable ex$) {
                throw new AssertionError("should not reach here", ex$);
            }
        };
    }
}
```
https://github.com/purejava/appindicator-gtk3-java/blob/ac5a6944eec0da3237e64dde150d37268051a978/src/main/java/org/purejava/appindicator/GCallback.java#L15

Reply:
> `jextract` generates a `SegmentScope`

Not sure if I really got the point of what you're saying, but what `jextract` has generated are methods that you need to feed in a scope, so Panama knows about the intended lifecycle of the `GCallback`. In other words: The scope must outlive the GCallback.

With it being "auto" it is my understanding that memory bound to it will be deallocated as soon as the last Java object referencing it has been GC'ed. Now if the _native_ code still wants to access the memory, it may run into a segfault.

If we use either the global scope (will live till the JVM terminates) or the scope of an Arena whose lifecycle we can explicitly control, we are in control of the GCallbacks end of life, rather than the GC.

Reply:
> Not sure if I really got the point of what you're saying, but what `jextract` has generated are methods that you need to feed in a scope, so Panama knows about the intended lifecycle of the `GCallback`. In other words: The scope must outlive the GCallback.
> 

My mentioning of the generated `SegmentScope` can be safely ignored. 😃 

> With it being "auto" it is my naive understanding that memory bound to it will be deallocated as soon as the last Java object referencing it has been GC'ed. Now if the _native_ code still wants to access the memory, it may run into a segfault.
> 

Yes, that's true. My experience with WinSparkle was, that GC happens very quickly, which led to my assumption, that the callbacks here survive GC, as the callbacks get executed. But that is guessing and no guarantee. 

> If we use either the global scope (will live till the JVM terminates) or the scope of an Arena whose lifecycle we can explicitly control, we are in control of the GCallbacks end of life, rather than the GC.
> 

So you are right, that we should use `Arena` here, whose lifecycle we can control.

Reply:
When we use the `Arena` like this:
```java
private static final Arena ARENA = Arena.openShared();
```
there's no way to `close()` the `Arena` once Cryptomator ends, right? Doesn't that lead to memory leakage?

I tried it with a try ressource block / `AutoClosable`, but that crahes the VM.

Reply:
Well if the memory should remain allocated as long as the jvm runs, I guess the global scope is appropriate.

Reply:
> I tried it with a try ressource block / `AutoClosable`, but that crahes the VM.

The try ressource block is ok, it workes with a confined `Arena`. The JVM crashed, because the callback needs to be allocated with a global `SegmentScope`. Otherwise it's gc'ed when executed.